### ISSUE TITLE
onnx export tool is missing from this fork.

### ISSUE THREAD

**Issue by SkybuckFlying** [JackismyShephard/ultimate-rvc#222]
G:\Tools\Ultimate-RVC\infer\modules\onnx\export.py

I recommend adding it so the obs-rvc plugin can be used.

This script is needed to convert rvc models/pth to onnx.

**Comment by SkybuckFlying** [comment]
Grok feedback to try:

The issue of "nothing being produced" when running your batch file to export an ONNX model could stem from several potential causes. Below, I’ll analyze the provided Python script and batch file, identify possible reasons for the failure, and suggest steps to diagnose and resolve the issue.

---

### Key Components of the Problem
1. **Python Script**: The script (`export_onnx`) is designed to load a PyTorch model (`.pth` file), convert it to ONNX format, and save the resulting `.onnx` file using `torch.onnx.export`. It also applies ONNX simplification using `onnxsim`.
2. **Batch File**: The batch file sets the `PYTHONPATH` and runs the Python script with specific arguments for the model path and output path.
3. **Symptom**: The script runs without producing the expected `.onnx` file, and no clear error messages are provided (or they are not being captured).

---

### Potential Reasons for "Nothing Being Produced"
Here are the likely reasons why the script is not producing the expected `.onnx` file:

#### 1. **Error in Python Script Execution**
   - **Issue**: The Python script might be failing silently due to an exception (e.g., file not found, module import error, or invalid model configuration).
   - **Why**: The batch file does not redirect Python errors to a log file, and the `echo Exit code: %ERRORLEVEL%` might not provide enough context if the script crashes silently.
   - **Diagnosis**:
     - Check if the Python script is running at all. Add `print` statements or logging to the `export_onnx` function to confirm it starts and progresses.
     - Modify the batch file to capture Python output and errors. For example:
       ```batch
       @echo on
       set PYTHONPATH=G:\Tools\Ultimate-RVC (Chinese)
       python -m infer.modules.onnx.export --model "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime_550e_15950s.pth" --output "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime.onnx" > output.log 2>&1
       echo Exit code: %ERRORLEVEL%
       pause
       ```
       This redirects both standard output and errors to `output.log`. Check this file for error messages.
   - **Common Errors**:
     - **FileNotFoundError**: Ensure the model file (`OptimusPrime_550e_15950s.pth`) exists at the specified path.
     - **ModuleNotFoundError**: Verify that the `infer.lib.infer_pack.models_onnx` module and its dependencies (`torch`, `onnx`, `onnxsim`) are installed in the Python environment.
     - **Invalid Model**: The `.pth` file might be corrupted or incompatible with the `SynthesizerTrnMsNSFsidM` model architecture.

#### 2. **Incorrect `PYTHONPATH` or Environment Issues**
   - **Issue**: The `PYTHONPATH` is set to `G:\Tools\Ultimate-RVC (Chinese)`, which may not include the necessary modules or may point to an incorrect directory.
   - **Why**: If the `infer` module or its dependencies are not in the `PYTHONPATH`, Python will fail to import `infer.lib.infer_pack.models_onnx`.
   - **Diagnosis**:
     - Verify that the directory `G:\Tools\Ultimate-RVC (Chinese)` contains the `infer` module or is the correct root for your project.
     - Run the Python command manually in a command prompt to see if the module imports correctly:
       ```cmd
       set PYTHONPATH=G:\Tools\Ultimate-RVC (Chinese)
       python -c "from infer.lib.infer_pack.models_onnx import SynthesizerTrnMsNSFsidM"
       ```
       If this fails, you’ll see a `ModuleNotFoundError`.
   - **Fix**:
     - Ensure all required packages (`torch`, `onnx`, `onnxsim`) are installed. Run:
       ```cmd
       pip install torch onnx onnxsim
       ```
     - If the `infer` module is part of a custom project, ensure the `PYTHONPATH` points to the correct directory containing the `infer` module.
     - Consider using a virtual environment to isolate dependencies and avoid conflicts.

#### 3. **Command-Line Argument Parsing Issue**
   - **Issue**: The script is invoked with `python -m infer.modules.onnx.export --model ... --output ...`, but the script itself doesn’t show how it handles these arguments.
   - **Why**: If the `export_onnx` function is not called correctly or if the command-line arguments are not parsed, the function might not execute.
   - **Diagnosis**:
     - Check if the `infer.modules.onnx.export` module has a `main` function or argument parser (e.g., using `argparse`) that calls `export_onnx(ModelPath, ExportedPath)`.
     - Add a minimal test script to verify argument passing:
       ```python
       import sys
       print("Arguments:", sys.argv)
       ```
       Run it with:
       ```cmd
       python -m infer.modules.onnx.export --model test.pth --output test.onnx
       ```
       Check if the arguments are received correctly.
   - **Fix**:
     - Ensure the `infer.modules.onnx.export` module includes code to parse `--model` and `--output` arguments and pass them to `export_onnx`. For example:
       ```python
       import argparse

       def main():
           parser = argparse.ArgumentParser()
           parser.add_argument("--model", required=True)
           parser.add_argument("--output", required=True)
           args = parser.parse_args()
           export_onnx(args.model, args.output)

       if __name__ == "__main__":
           main()
       ```

#### 4. **ONNX Export Failure**
   - **Issue**: The `torch.onnx.export` call might fail due to incompatible model operations, incorrect input shapes, or issues with the model architecture.
   - **Why**: Not all PyTorch operations are supported by ONNX, especially for complex models like `SynthesizerTrnMsNSFsidM`. Additionally, the input tensors (`test_phone`, `test_phone_lengths`, etc.) might not match the model’s expectations.
   - **Diagnosis**:
     - Add error handling around the `torch.onnx.export` call:
       ```python
       try:
           torch.onnx.export(
               net_g,
               (
                   test_phone.to(device),
                   test_phone_lengths.to(device),
                   test_pitch.to(device),
                   test_pitchf.to(device),
                   test_ds.to(device),
                   test_rnd.to(device),
               ),
               ExportedPath,
               dynamic_axes={
                   "phone": [1],
                   "pitch": [1],
                   "pitchf": [1],
                   "rnd": [2],
               },
               do_constant_folding=False,
               opset_version=18,
               verbose=True,  # Enable verbose to debug
               input_names=input_names,
               output_names=output_names,
           )
           print("ONNX export successful")
       except Exception as e:
           print(f"ONNX export failed: {e}")
       ```
     - Set `verbose=True` in `torch.onnx.export` to get detailed logs about the export process.
     - Verify that the input tensor shapes (`test_phone`, `test_pitch`, etc.) match the model’s expected input shapes. For example, ensure `vec_channels` (256 or 768) aligns with the model’s configuration.
   - **Fix**:
     - If specific operations are unsupported, simplify the model or use a different `opset_version` (e.g., try `opset_version=16` or `17`).
     - Ensure the model is in evaluation mode before export:
       ```python
       net_g.eval()
       ```
     - Check the model’s forward pass with the test inputs to ensure it runs without errors:
       ```python
       with torch.no_grad():
           output = net_g(
               test_phone.to(device),
               test_phone_lengths.to(device),
               test_pitch.to(device),
               test_pitchf.to(device),
               test_ds.to(device),
               test_rnd.to(device),
           )
           print("Model output shape:", output.shape)
       ```

#### 5. **ONNX Simplification Failure**
   - **Issue**: The `onnxsim.simplify` step might fail, causing the script to exit without saving the final `.onnx` file.
   - **Why**: The ONNX model might be too complex, or `onnxsim` might encounter unsupported operations.
   - **Diagnosis**:
     - Add error handling around the simplification step:
       ```python
       try:
           model, _ = onnxsim.simplify(ExportedPath)
           onnx.save(model, ExportedPath)
           print("ONNX simplification successful")
       except Exception as e:
           print(f"ONNX simplification failed: {e}")
       ```
     - Try running the script without simplification to isolate the issue:
       ```python
       # Comment out simplification
       # model, _ = onnxsim.simplify(ExportedPath)
       # onnx.save(model, ExportedPath)
       onnx.save(onnx.load(ExportedPath), ExportedPath)  # Save without simplification
       ```
   - **Fix**:
     - If simplification fails, skip it and use the unsimplified `.onnx` file for testing.
     - Update `onnxsim` to the latest version:
       ```cmd
       pip install --upgrade onnxsim
       ```

#### 6. **File Path Issues**
   - **Issue**: The input model path or output path might be incorrect or inaccessible.
   - **Why**: Windows paths with spaces (e.g., `Transformers - Optimus Prime`) can cause issues if not handled properly, or the script might lack write permissions for the output directory.
   - **Diagnosis**:
     - Verify that the input file exists:
       ```cmd
       dir "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime_550e_15950s.pth"
       ```
     - Verify that the output directory is writable by creating a test file:
       ```cmd
       echo test > "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\test.txt"
       ```
     - Check if the paths are correctly passed to the script by printing them:
       ```python
       print(f"ModelPath: {ModelPath}")
       print(f"ExportedPath: {ExportedPath}")
       ```
   - **Fix**:
     - Use raw strings or double backslashes for Windows paths in the batch file:
       ```batch
       python -m infer.modules.onnx.export --model "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime_550e_15950s.pth" --output "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime.onnx"
       ```
     - Ensure the script has write permissions for the output directory.

#### 7. **Version or Dependency Mismatch**
   - **Issue**: The versions of `torch`, `onnx`, or `onnxsim` might be incompatible with the model or the ONNX export process.
   - **Why**: The script uses `opset_version=18`, which requires a recent version of PyTorch and ONNX. Older versions might not support this opset or the model’s operations.
   - **Diagnosis**:
     - Check the installed versions:
       ```cmd
       pip show torch onnx onnxsim
       ```
     - Ensure PyTorch supports ONNX opset 18 (requires PyTorch 1.11 or later).
   - **Fix**:
     - Update dependencies:
       ```cmd
       pip install --upgrade torch onnx onnxsim
       ```
     - If issues persist, try a lower opset version (e.g., `opset_version=16`).

---

### Debugging Steps
To systematically diagnose the issue, follow these steps:

1. **Capture Output**:
   - Modify the batch file to log output and errors (see modified batch file above).
   - Check `output.log` for any error messages.

2. **Test Python Environment**:
   - Verify that all required modules are installed and accessible:
     ```cmd
     python -c "import torch, onnx, onnxsim, infer.lib.infer_pack.models_onnx"
     ```
   - If this fails, fix the `PYTHONPATH` or install missing dependencies.

3. **Run Script Interactively**:
   - Open a Python interpreter and run the `export_onnx` function manually:
     ```python
     from infer.modules.onnx.export import export_onnx
     export_onnx(
         "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime_550e_15950s.pth",
         "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime.onnx"
     )
     ```
   - Add `print` statements to track progress.

4. **Isolate ONNX Export**:
   - Comment out the `onnxsim.simplify` step and check if the `.onnx` file is created before simplification.
   - If the file is created, the issue is with simplification.

5. **Check Model and Inputs**:
   - Verify that the `.pth` file loads correctly:
     ```python
     import torch
     cpt = torch.load("G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime_550e_15950s.pth", map_location="cpu")
     print(cpt.keys())
     ```
   - Ensure the input tensors match the model’s expectations.

---

### Suggested Fixes
Based on the likely issues, here are immediate fixes to try:

1. **Add Error Handling to the Script**:
   ```python
   import torch
   import onnxsim
   import onnx
   from infer.lib.infer_pack.models_onnx import SynthesizerTrnMsNSFsidM

   def export_onnx(ModelPath, ExportedPath):
       try:
           print(f"Loading model from {ModelPath}")
           cpt = torch.load(ModelPath, map_location="cpu")
           print("Model loaded successfully")
           cpt["config"][-3] = cpt["weight"]["emb_g.weight"].shape[0]
           vec_channels = 256 if cpt.get("version", "v1") == "v1" else 768

           test_phone = torch.rand(1, 200, vec_channels)
           test_phone_lengths = torch.tensor([200]).long()
           test_pitch = torch.randint(size=(1, 200), low=5, high=255)
           test_pitchf = torch.rand(1, 200)
           test_ds = torch.LongTensor([0])
           test_rnd = torch.rand(1, 192, 200)

           device = "cpu"
           net_g = SynthesizerTrnMsNSFsidM(
               *cpt["config"], is_half=False, version=cpt.get("version", "v1")
           )
           net_g.load_state_dict(cpt["weight"], strict=False)
           net_g.eval()
           print("Model initialized successfully")

           input_names = ["phone", "phone_lengths", "pitch", "pitchf", "ds", "rnd"]
           output_names = ["audio"]

           print(f"Exporting to {ExportedPath}")
           torch.onnx.export(
               net_g,
               (
                   test_phone.to(device),
                   test_phone_lengths.to(device),
                   test_pitch.to(device),
                   test_pitchf.to(device),
                   test_ds.to(device),
                   test_rnd.to(device),
               ),
               ExportedPath,
               dynamic_axes={
                   "phone": [1],
                   "pitch": [1],
                   "pitchf": [1],
                   "rnd": [2],
               },
               do_constant_folding=False,
               opset_version=18,
               verbose=True,
               input_names=input_names,
               output_names=output_names,
           )
           print("ONNX export completed")

           print("Simplifying ONNX model")
           model, _ = onnxsim.simplify(ExportedPath)
           onnx.save(model, ExportedPath)
           print("ONNX simplification completed")
           return "Finished"
       except Exception as e:
           print(f"Error: {e}")
           raise
   ```

2. **Update the Batch File**:
   ```batch
   @echo on
   set PYTHONPATH=G:\Tools\Ultimate-RVC (Chinese)
   python -m infer.modules.onnx.export --model "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime_550e_15950s.pth" --output "G:\Tools\Ultimate-RVC (JackismyShephard)\models\rvc\voice_models\Transformers - Optimus Prime\OptimusPrime.onnx" > output.log 2>&1
   echo Exit code: %ERRORLEVEL%
   type output.log
   pause
   ```

3. **Verify Dependencies**:
   ```cmd
   pip install --upgrade torch onnx onnxsim
   ```

4. **Test Without Simplification**:
   - Temporarily comment out the `onnxsim.simplify` and `onnx.save` lines to see if the raw ONNX file is created.

---

### Expected Outcome
After applying these fixes, you should see:
- The `output.log` file containing detailed logs or error messages.
- If successful, the `OptimusPrime.onnx` file in the specified output directory.
- If an error occurs, the exact exception message will help pinpoint the issue (e.g., file not found, unsupported operation, or import error).

If the issue persists after trying these steps, please share the contents of `output.log` or any error messages, and I can provide further assistance.

**Comment by SkybuckFlying** [comment]
Nope, still crashes...

Adding crash file, maybe it will help somebody or the AI, yes AI can read it... going to link this in chinese issue, maybe they can follow along, don't know.

[Crash 2025-10-20 13-50-19.txt](https://github.com/user-attachments/files/23000693/Crash.2025-10-20.13-50-19.txt)
